### DISCUSSION TITLE
Using symmetries effectively to get lowest 10 energies for long chains (Anisotropic XY Model)

### DISCUSSION THREAD

**Discussion by amoyerjlab** [QuSpin/QuSpin#488] [Q&A]
I am studying the Anisotropic XY 1D spin chain. I am supposed to calculate the lowest 10 energies for up to chain lengths of 43 (or as high as I possibly can.) I can calculate the lowest 10 energies without symmetries up to L=27 on our computer. After this, the basis size exceeds the computers memory capacity. I have decided to do the calculations in different symmetry blocks. So if I implement the Nup symmetry, QuSpin takes the original basis and breaks it into L+1 complementary bases. The largest of these is the L/2 subspace. But after implementing this symmetry, the L/2 subspace for L=30 exceeds the entire size of the Hilbert space for L=27. So If I only implement this symmetry, I can theoretically only calculate the lowest 10 energies for L=28 and L=29.

I picked the Nup symmetry to start with because all the subspaces combine to for the entire hilbert space. I am trying to use other symmetries, but the code will get messier. For example, other symmetries may not divide up the states so cleanly. After calculating the eigenvalues for the subspaces, it will be tricky to reassemble the true lowest 10 energies because some states may have been omitted and some may end up in multiple blocks. Can anyone offer any insight on how to do this analysis cleanly?

This is my code for only using the Nup Symmetry: 

def AnisotropicXY(gamma=0.5,length=4):
   
    L = length
    g = gamma
    j_x = (1+g)
    j_y = (1-g)
    
    Lx, Ly = L, 1 # linear dimension of spin 1 2d lattice
    N_2d = Lx*Ly # number of sites for spin 1
    
    ###### setting up user-defined symmetry transformations for 2d lattice ######
    s = np.arange(N_2d) # sites [0,1,2,....]
    x = s%Lx # x positions for sites
    y = s//Lx # y positions for sites
    T_x = (x+1)%Lx + Lx*y # translation along x-direction
    T_y = x +Lx*((y+1)%Ly) # translation along y-direction
    P_x = x + Lx*(Ly-y-1) # reflection about x-axis
    P_y = (Lx-x-1) + Lx*y # reflection about y-axis
    Z   = -(s+1) # spin inversion
    
    # compute spin-1/2 basis
    #basis = spin_basis_general(N_2d,pauli=False,Nup=L,kxblock=(T_x,0),pxblock=(P_x,0),zblock=(Z,0))
    basis = spin_basis_general(L,pauli=False,Nup=int(L/2))
    
    print(basis.Ns)
    
    # define operators with PBC using site-coupling lists
    J_X = [[j_x,i,(i+1)%L] for i in range(L)] # PBC
    J_Y = [[j_y,i,(i+1)%L] for i in range(L)] # PBC
    static = [["xx",J_X],["yy",J_Y]] # Pick the type of spin operator for the coefficients
    dynamic=[]
    
    # compute the time-dependent Heisenberg Hamiltonian
    H_XY = hamiltonian(static,dynamic,basis=basis,dtype=np.float64,check_symm=False,check_pcon=False)
    
    #Calculate energy spectrum
    #E=eigsh(H_XY.aslinearoperator(),k=11,which='SA',return_eigenvectors=False)
    #E=E.tolist()
    #E = sorted(E, key = float)
    #E.append(gamma)
    
    return None

Thank you for your time,
Andrew

**Comment by JiahaoYao** [comment]
Hi @amoyerjlab, can I understand that your issue is that when implementing other symmetries, the lowest 10 energies will end up in different sectors? 

by the way, if the spin system is 1d, u can also use [spin basis 1d](http://weinbe58.github.io/QuSpin/generated/quspin.basis.spin_basis_1d.html#quspin.basis.spin_basis_1d). It will be easier to add different symmetries.

**Reply by amoyerjlab** [reply]
您好 JiahaoYao. Thank you for looking at my post. You are correct that my issue is how to get the 10 lowest energies after doing calculations in different sectors. I think I know how to do it if I only used the Nup symmetry. But adding the new symmetries will make it more complicated. I am just looking to see if anyone has any experience with this issue or advice on how to structure my code. I know I can use spin_basis_1d, but this function does not use parallel processing. That is why I am using the spin_basis_general.

**Comment by mgbukov** [comment]
@amoyerjlab your best shot is to use as many symmetries as you can, so you can access large system sizes. Then you can study which symmetry sectors the lowest 10 energies are distributed in, for small system sizes [such a pattern will likely vary due to even-odd effects]; finally, you can try to extrapolate your knowledge to the larger system sizes (if a pattern exists). I doubt you will be able to reach 43 spins, but if you do, please let us know how it went.

For `g=0` you should be able to apply a global z-rotation to your Hamiltonian which removes the `yy` term and aligns the interaction along `xx` with some effective couling, but then you have a classical Ising model and you can find all energies easily. 

Btw, the XY model in 1d may be mappable to free fermions using a JW transformation; if that's the case and you can get an exact analytical solution, you won't need quspin.
